Maybe sync.Pool somehow causes panic

running below code snippet has chance of causing SIGSEGV.
Either comment out the avro.Unmarshal() or log.Info() would prevent segmentation fault, which is very possibly a memory problem...


```
package main

import (
	"fmt"
	"github.com/rs/zerolog"
	"os"
	"strings"
	"time"

	"github.com/hamba/avro"
	"github.com/rs/zerolog/log"
)

func main() {
	// Use zerolog console writer
	output := zerolog.ConsoleWriter{Out: os.Stderr, NoColor: false, TimeFormat: time.RFC3339}

	output.FormatLevel = func(i interface{}) string {
		return strings.ToUpper(fmt.Sprintf("| %-6s|", i))
	}
	output.FormatMessage = func(i interface{}) string {
		return fmt.Sprintf("%s", i)
	}
	output.FormatFieldName = func(i interface{}) string {
		return fmt.Sprintf("%s=", i)
	}
	output.FormatFieldValue = func(i interface{}) string {
		return strings.ToUpper(fmt.Sprintf("%s", i))
	}

	log.Logger = zerolog.New(output).With().Timestamp().Logger().With().Caller().Logger()

	// Begin our test
	type UserLabelMeta struct {
		Language string `avro:"language"`
		Ip       string `avro:"ip"`
	}
	type Event struct {
		UserId    string `avro:"user_id"`
		Timestamp int64  `avro:"timestamp"`
		// User related
		UserLabels map[string]string `avro:"user_labels"`
		MetaLabels *UserLabelMeta    `avro:"meta_labels"`
	}

	in := Event{
		UserId:    "UserId",
		Timestamp: int64(time.Now().UnixNano() / 1e6),

		UserLabels: map[string]string{"userLabel1": "userValue1"},

		MetaLabels: &UserLabelMeta{
			Language: "English",
			Ip:       "1.1.1.1",
		},
	}
	// Init schema
	schema, err := avro.Parse(string(`{
  "namespace": "minerva.dlive.tv",
  "type": "record",
  "name": "Event",
  "fields": [
    {
      "name": "user_id", "type":  "string"
    },
    {
      "name": "timestamp", "type": "long"
    },
    {
      "name": "user_labels",
      "type": {
        "type": "map",
        "values": "string"
      }
    },
    {
      "name": "meta_labels",
      "type": ["null", {
        "type": "record",
        "name": "UserLabelMeta",
        "fields": [
          { "name": "language", "type": "string" },
          { "name": "ip", "type": "string" }
        ]
      }],
      "default": null
    }
  ]
}`))
	if err != nil {
		panic(err)
	}

	// marshal
	data, err := avro.Marshal(schema, in)
	if err != nil {
		panic(err)
	}
	log.Info().Msgf("avro.Marshal(in)=% 08x", data)

	// unmarshal
	out := Event{}
	err = avro.Unmarshal(schema, data, &out)
	if err != nil {
		panic(err)
	}

	log.Info().Msgf("possibly cause SIGSEGV")

}
```

**Log output:**
```
2019-06-14T17:32:32-07:00 | INFO  | cmd/tmp/main.go:99 > avro.Marshal(in)=0c 55 73 65 72 49 64 ca ce e6 88 eb 5a 01 2c 14 75 73 65 72 4c 61 62 65 6c 31 14 75 73 65 72 56 61 6c 75 65 31 00 02 0e 45 6e 67 6c 69 73 68 0e 31 2e 31 2e 31 2e 31
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0xf pc=0x100db95]

goroutine 1 [running]:
reflect.mapassign(0x11f72e0, 0x7, 0xc000013610, 0xc0000135f0)
    /usr/local/Cellar/go/1.12.5/libexec/src/runtime/map.go:1331 +0x3f
reflect.Value.SetMapIndex(0x11f72e0, 0xc0000103a8, 0x195, 0x11e51e0, 0xc000013610, 0x98, 0x11f6800, 0xc0000135f0, 0x194)
    /usr/local/Cellar/go/1.12.5/libexec/src/reflect/value.go:1615 +0x234
encoding/json.(*decodeState).object(0xc000014188, 0x11dcfa0, 0xc0000103a8, 0x16, 0xc0000141b0, 0xc00001a77b)
    /usr/local/Cellar/go/1.12.5/libexec/src/encoding/json/decode.go:805 +0xd97
encoding/json.(*decodeState).value(0xc000014188, 0x11dcfa0, 0xc0000103a8, 0x16, 0x20, 0xa)
    /usr/local/Cellar/go/1.12.5/libexec/src/encoding/json/decode.go:381 +0x6e
encoding/json.(*decodeState).unmarshal(0xc000014188, 0x11dcfa0, 0xc0000103a8, 0x0, 0xc000114001)
    /usr/local/Cellar/go/1.12.5/libexec/src/encoding/json/decode.go:179 +0x209
encoding/json.(*Decoder).Decode(0xc000014160, 0x11dcfa0, 0xc0000103a8, 0xc0000b8060, 0x1409c60)
    /usr/local/Cellar/go/1.12.5/libexec/src/encoding/json/stream.go:73 +0x187
github.com/rs/zerolog.ConsoleWriter.Write(0x1278800, 0xc000010018, 0x0, 0x123cd82, 0x19, 0xc00006c840, 0x4, 0x4, 0x0, 0x1243990, ...)
    /Users/xin/Developer/go/pkg/mod/github.com/rs/zerolog@v1.14.3/console.go:101 +0x186
github.com/rs/zerolog.levelWriterAdapter.WriteLevel(...)
    /Users/xin/Developer/go/pkg/mod/github.com/rs/zerolog@v1.14.3/writer.go:20
github.com/rs/zerolog.(*Event).write(0xc000024240, 0x9f, 0x1f4)
    /Users/xin/Developer/go/pkg/mod/github.com/rs/zerolog@v1.14.3/event.go:75 +0x125
github.com/rs/zerolog.(*Event).msg(0xc000024240, 0xc000018300, 0x2f)
    /Users/xin/Developer/go/pkg/mod/github.com/rs/zerolog@v1.14.3/event.go:134 +0x13d
github.com/rs/zerolog.(*Event).Msgf(0xc000024240, 0x1241e3c, 0x2f, 0x0, 0x0, 0x0)
    /Users/xin/Developer/go/pkg/mod/github.com/rs/zerolog@v1.14.3/event.go:116 +0x83
main.main()
    /Users/xin/Developer/go/src/github.com/lino-network/minerva/cmd/tmp/main.go:108 +0xb5e
```

--- a/codec_union.go
+++ b/codec_union.go
@@ -146,18 +146,19 @@ func (e *mapUnionEncoder) Encode(ptr unsafe.Pointer, w *Writer) {
 func decoderOfPtrUnion(cfg *frozenConfig, schema Schema, typ reflect2.Type) ValDecoder {
 	union := schema.(*UnionSchema)
 	ptrType := typ.(*reflect2.UnsafePtrType)
-	decoder := decoderOfType(cfg, union.Types()[1], ptrType.Elem())
+	elemType := ptrType.Elem()
+	decoder := decoderOfType(cfg, union.Types()[1], elemType)
 
 	return &unionPtrDecoder{
 		schema:  union,
-		typ:     ptrType,
+		typ:     elemType,
 		decoder: decoder,
 	}
 }
 
 type unionPtrDecoder struct {
 	schema  *UnionSchema
-	typ     *reflect2.UnsafePtrType
+	typ     reflect2.Type
 	decoder ValDecoder
 }
 
@@ -172,9 +173,16 @@ func (d *unionPtrDecoder) Decode(ptr unsafe.Pointer, r *Reader) {
 		return
 	}
 
-	newPtr := d.typ.UnsafeNew()
-	d.decoder.Decode(newPtr, r)
-	*((*unsafe.Pointer)(ptr)) = newPtr
+	if *((*unsafe.Pointer)(ptr)) == nil {
+		// Create new instance
+		newPtr := d.typ.UnsafeNew()
+		d.decoder.Decode(newPtr, r)
+		*((*unsafe.Pointer)(ptr)) = newPtr
+		return
+	}
+
+	// Reuse existing instance
+	d.decoder.Decode(*((*unsafe.Pointer)(ptr)), r)
 }
 
 func encoderOfPtrUnion(cfg *frozenConfig, schema Schema, typ reflect2.Type) ValEncoder {
@@ -184,14 +192,12 @@ func encoderOfPtrUnion(cfg *frozenConfig, schema Schema, typ reflect2.Type) ValE
 
 	return &unionPtrEncoder{
 		schema:  union,
-		typ:     ptrType,
 		encoder: encoder,
 	}
 }
 
 type unionPtrEncoder struct {
 	schema  *UnionSchema
-	typ     *reflect2.UnsafePtrType
 	encoder ValEncoder
 }
 

--- a/decoder_union_test.go
+++ b/decoder_union_test.go
@@ -105,6 +105,24 @@ func TestDecoder_UnionPtr(t *testing.T) {
 	assert.Equal(t, &want, got)
 }
 
+func TestDecoder_UnionPtrReuseInstance(t *testing.T) {
+	defer ConfigTeardown()
+
+	avro.Register("test", &TestRecord{})
+
+	data := []byte{0x02, 0x36, 0x06, 0x66, 0x6F, 0x6F}
+	schema := `["null", {"type": "record", "name": "test", "fields" : [{"name": "a", "type": "long"}, {"name": "b", "type": "string"}]}]`
+	dec, _ := avro.NewDecoder(schema, bytes.NewReader(data))
+
+	got := &TestRecord{}
+	err := dec.Decode(&got)
+
+	assert.NoError(t, err)
+	assert.IsType(t, &TestRecord{}, got)
+	assert.Equal(t, int64(27), got.A)
+	assert.Equal(t, "foo", got.B)
+}
+
 func TestDecoder_UnionPtrNull(t *testing.T) {
 	defer ConfigTeardown()
 

